ledger/participant-integration-api/src/main/scala/platform/indexer/MeteringAggregator.scala

```diff
@@ -0,0 +1,180 @@
+              ()
+            }
+          },
+          period.toMillis,
+          period.toMillis,
+        )
+      }
+    }
+  }
+
+  private def toTimestamp(dateTime: OffsetDateTime): Timestamp =
+    Timestamp.assertFromInstant(dateTime.toInstant)
+  private def toOffsetDateTime(timestamp: Timestamp): OffsetDateTime =
+    OffsetDateTime.ofInstant(timestamp.toInstant, ZoneOffset.UTC)
+
+}
+
+class MeteringAggregator(
+    meteringStore: MeteringStorageWriteBackend,
+    parameterStore: ParameterStorageBackend,
+    metrics: Metrics,
+    clock: () => Timestamp = () => Timestamp.now(),
+)(implicit loggingContext: LoggingContext) {
+
+  private implicit val ec: ExecutionContext = ExecutionContext.global
+
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private[platform] def initialize(sqlExecutor: SqlExecutor): Future[Unit] = {
+    val initTimestamp = toOffsetDateTime(clock()).truncatedTo(ChronoUnit.HOURS).minusHours(1)
+    val initLedgerMeteringEnd = LedgerMeteringEnd(Offset.beforeBegin, toTimestamp(initTimestamp))
+    sqlExecutor.executeSql(metrics.daml.index.db.initializeMeteringAggregator) {
+      parameterStore.initializeLedgerMeteringEnd(initLedgerMeteringEnd)
+    }
+  }
+
+  private[platform] def run(sqlExecutor: SqlExecutor): Future[Unit] = {
+
+    val future = sqlExecutor.executeSql(metrics.daml.index.db.meteringAggregator) { conn =>
+      val nowUtcTime = toOffsetDateTime(clock())
+      val lastLedgerMeteringEnd = getLedgerMeteringEnd(conn)
+      val startUtcTime: OffsetDateTime = toOffsetDateTime(lastLedgerMeteringEnd.timestamp)
+      val endUtcTime = startUtcTime.plusHours(1)
+
+      if (nowUtcTime.isAfter(endUtcTime)) {
+
+        val toEndTime = toTimestamp(endUtcTime)
+        val ingestedLedgerEnd = parameterStore.ledgerEnd(conn).lastOffset
+        val maybeMaxOffset =
+          meteringStore.transactionMeteringMaxOffset(lastLedgerMeteringEnd.offset, toEndTime)(conn)
+
+        val (
+          periodIngested, // This is true if the time period is closed fully ingested
+          hasMetering, // This is true if there are transaction_metering records
+          toOffsetEnd, // This is the 'to' offset for the period being aggregated
+        ) = maybeMaxOffset match {
+          case Some(offset) => (offset <= ingestedLedgerEnd, true, offset)
+          case None => (true, false, lastLedgerMeteringEnd.offset)
+        }
+
+        if (periodIngested) {
+          Some(
+            aggregate(
+              conn = conn,
+              lastLedgerMeteringEnd = lastLedgerMeteringEnd,
+              thisLedgerMeteringEnd = LedgerMeteringEnd(toOffsetEnd, toEndTime),
+              hasMetering = hasMetering,
+            )
+          )
+        } else {
+          logger.info("Not all transaction metering for aggregation time period is yet ingested")
+          None
+        }
+      } else {
+        None
+      }
+    }
+
+    future.onComplete({
```

Comment:
having an option result above only that we do logging based on that here seems a little bit unnecessary to me. Would you consider adding back the logging where and not doing onComplete at all? (that would solve the execution context need as well ;) )

Reply:
As discussed the benefit of this approach is that it capture problems with failed futures.  Also it will only print success of the future successfully completing which is possible reassuring.  Have switched to use `parasitic` context.